### ISSUE TITLE
Can we optimize non-locking RMW atomic operations?

### ISSUE THREAD

**Issue by Sonicadvance1** [FEX-Emu/FEX#1729]
Currently we convert all lock RMW ops to acquire-release semantics.

Couple weird things to investigate here

1. Basic ALU ops without lock
    - Non-lock ops get turned in to load + ALU + store
    - Can potentially convert in to atomic memory operation **without** acquire-release semantics.
    - Should only generate on ARMv8.1+ if it supports atomic memory ops
 2. RMW ops that don't imply LOCK but really should, used without LOCK
    - **CMPXCHG, CMPXCHG8B, CMPXCHG16B, XADD**
    - These instructions don't imply LOCK prefixes but they are almost universally used with them
    - Linux kernel has some optimization where it backpatches `lock cmpxchg` in to `nop cmpxchg` on uniprocessors? Citation needed.
    - These might be able to be converted to operations with...release? semantics?
    - Needs investigation.

**Comment by dnadlinger** [comment]
> Linux kernel has some optimization where it backpatches lock cmpxchg in to nop cmpxchg on uniprocessors? Citation needed.

LOCK_PREFIX is defined here

https://github.com/torvalds/linux/blob/8291eaafed36f575f23951f3ce18407f480e9ecf/arch/x86/include/asm/alternative.h#L16-L50

and the patching mechanism is

https://github.com/torvalds/linux/blob/8291eaafed36f575f23951f3ce18407f480e9ecf/arch/x86/kernel/alternative.c#L264

**Comment by Sonicadvance1** [comment]
Nah, let's not go here. Single threaded applications already are pretty locked down and games aren't ever single threaded so any gains aren't worth it.
